commons/src/main/java/org/infinispan/multimap/api/MultimapCache.java

```diff
@@ -0,0 +1,91 @@
+package org.infinispan.multimap.api;
+
+import java.util.Collection;
+import java.util.concurrent.CompletableFuture;
+import java.util.function.Predicate;
+
+/**
+ * MultimapCache provides the common building block for the two different types of multimap caches that Infinispan
+ * provides: embedded and remote.
+ * <p>
+ * Please see the <a href="http://www.jboss.org/infinispan/docs">Infinispan documentation</a> and/or the <a
+ * href="https://docs.jboss.org/author/display/ISPN/Getting+Started+Guide#GettingStartedGuide-5minutetutorial">5 Minute
+ * Usage Tutorial</a> for more details.
+ * <p/>
+ *
+ * @author Katia Aresti, karesti@redhat.com
+ * @see <a href="http://www.jboss.org/infinispan/docs">Infinispan documentation</a>
+ * @since 9.1
+ */
+public interface MultimapCache<K, V> {
+
+   /**
+    * Puts a value on a key.
+    * If the multimap cache supports duplicates, the value will be always added.
+    * If the multimap cache does not support duplicates and the value exists on the key, nothing will be done.
+    *
+    * @param key, the key to be put
+    * @param value, the value to added
+    */
+   CompletableFuture<Void> put(K key, V value);
+
+   /***
+    * Returns a view collection of the values associated with key in this multimap cache, if any.
+    * Any changes to the retrieved collection won't change the values in this multimap cache.
+    *
+    * @param key, to be retrieved
+    * @return the collection found in the key when it's present,
+    *         empty Collection when the key is not present
+    */
+   CompletableFuture<Collection<V>> get(K key);
+
+   /**
+    * Removes the key from the multimap cache, if such exists.
+    *
+    * @param key, to be removed
+    * @return true if the key was removed, false if not
+    */
+   CompletableFuture<Boolean> remove(K key);
+
+   /**
+    * Removes a value on a key. In the case where duplicates are supported, all the values will be removed.
+    * After remove call, the remaining value on the key might be an Empty Collection. In this case, the key is removed
+    * fro this multimap cache.
+    * Returns true when the value has been removed from the key.
+    *
+    * @param key, to be removed
+    * @param value, to be removed
+    * @return boolean, remove result
+    */
+   CompletableFuture<Boolean> remove(K key, V value);
+
+   /**
+    * Removes all the values in every key that matches the given {@link Predicate}.
```

Comment:
@danberindei it's not a the **key** that matches the predicate, since the predicate does not accept the key, therefore *values ... match*